File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        dp = [[] for _ in range(target + 1)]
        for c in candidates:
            for i in range(c, target + 1):
                if i == c: 
                    dp[i].append([c])
                    continue
                for com in dp[i - c]:
                    dp[i].append(com + [c])
        return dp[-1];

    # Does list concatenation with the `+` operator always return a new `list` instance?
    # https://stackoverflow.com/questions/64644671/does-list-concatenation-with-the-operator-always-return-a-new-list-instanc
    # The reference documentation does not explicitly guarantee that :list + :list results in a new list. However, it is implied by the += augmented assignment being explicitly in-place if possible in contrast to the + operator.

File: 39-combination-sum/combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<List<Integer>>> dp = new ArrayList<>();
        for (int i = 0; i <= target; i++) {
            dp.add(new ArrayList<>());
        }

        for (int c : candidates) {
            for (int i = c; i <= target; i++) {
                if (i == c) dp.get(i).add(Arrays.asList(c));
                else if (i > c) {
                    for (List<Integer> comb : dp.get(i - c)) {
                        List<Integer> tmp = new ArrayList<>();
                        tmp.addAll(comb);
                        tmp.add(c);
                        dp.get(i).add(tmp);
                    }
                }
            }
        }
        return dp.get(target);
    }
}